[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PolitoVandal/project-sim2real-rialti-giunti-gjinaj/blob/main/colab_template/test_random_policy.ipynb)

In [25]:
!rm -rf "project-sim2real-rialti-giunti-gjinaj"  # Cancella la cartella e il suo contenuto usando il comando !rm
!rm -rf "sim2real"  # Cancella la cartella e il suo contenuto usando il comando !rm
!rm -rf "sample_data"  # Cancella la cartella e il suo contenuto usando il comando !rm
!git clone https://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@github.com/PolitoVandal/project-sim2real-rialti-giunti-gjinaj
import os
import warnings

if not os.path.exists('sim2real/models'):
    os.makedirs('sim2real/models')

if not os.path.exists('sim2real/plots'):
    os.makedirs('sim2real/plots')

Cloning into 'project-sim2real-rialti-giunti-gjinaj'...
remote: Enumerating objects: 175, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 175 (delta 17), reused 18 (delta 12), pack-reused 144 (from 2)
Receiving objects: 100% (175/175), 959.67 KiB | 5.13 MiB/s, done.
Resolving deltas: 100% (90/90), done.


In [5]:
!apt-get install -y \
    libgl1-mesa-dev \
    libgl1-mesa-glx \
    libglew-dev \
    libosmesa6-dev \
    software-properties-common

!apt-get install -y patchelf

!pip install gym
!pip install free-mujoco-py
!pip install importlib-metadata
!pip install shimmy
!pip install stable-baselines3[extra]

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
software-properties-common is already the newest version (0.99.22.9).
The following additional packages will be installed:
  libegl-dev libgl-dev libgles-dev libgles1 libglu1-mesa libglu1-mesa-dev libglvnd-core-dev
  libglvnd-dev libglx-dev libopengl-dev libosmesa6
The following NEW packages will be installed:
  libegl-dev libgl-dev libgl1-mesa-dev libgl1-mesa-glx libgles-dev libgles1 libglew-dev
  libglu1-mesa libglu1-mesa-dev libglvnd-core-dev libglvnd-dev libglx-dev libopengl-dev libosmesa6
  libosmesa6-dev
0 upgraded, 15 newly installed, 0 to remove and 49 not upgraded.
Need to get 4,013 kB of archives.
After this operation, 19.4 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 libglx-dev amd64 1.4.0-1 [14.1 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgl-dev amd64 1.4.0-1 [101 kB]
Get:3 http://archive.ubuntu.com/ubuntu 

Set up the custom Hopper environment and provided util functions



1.   Upload `custom_hopper.zip` to the current session's file storage
2.   Un-zip it by running cell below


**TASK 1**


In [ ]:
import os
os.chdir('project-sim2real-rialti-giunti-gjinaj/colab_template')
!unzip custom_hopper.zip

Archive:  custom_hopper.zip
   creating: env/
  inflating: env/__init__.py         
  inflating: env/custom_hopper.py    
  inflating: env/mujoco_env.py       
   creating: env/assets/
  inflating: env/assets/hopper.xml   




---




**Test a random policy on the Gym Hopper environment**

\



Play around with this code to get familiar with the
Hopper environment.

For example, what happens if you don't reset the environment
even after the episode is over?
When exactly is the episode over?
What is an action here?

In [ ]:
import gym
from env.custom_hopper import *

/usr/local/lib/python3.10/dist-packages/Cython/Distutils/old_build_ext.py:15: DeprecationWarning: dep_util is Deprecated. Use functions from setuptools instead.
  from distutils.dep_util import newer, newer_group


Compiling /usr/local/lib/python3.10/dist-packages/mujoco_py/cymj.pyx because it changed.
[1/1] Cythonizing /usr/local/lib/python3.10/dist-packages/mujoco_py/cymj.pyx


INFO:root:running build_ext
INFO:root:building 'mujoco_py.cymj' extension
INFO:root:creating /usr/local/lib/python3.10/dist-packages/mujoco_py/generated/_pyxbld_2.0.2.13_310_linuxcpuextensionbuilder/temp.linux-x86_64-cpython-310/usr/local/lib/python3.10/dist-packages/mujoco_py
INFO:root:creating /usr/local/lib/python3.10/dist-packages/mujoco_py/generated/_pyxbld_2.0.2.13_310_linuxcpuextensionbuilder/temp.linux-x86_64-cpython-310/usr/local/lib/python3.10/dist-packages/mujoco_py/gl
INFO:root:x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/local/lib/python3.10/dist-packages/mujoco_py -I/usr/local/lib/python3.10/dist-packages/mujoco_py/binaries/linux/mujoco210/include -I/usr/local/lib/python3.10/dist-packages/numpy/core/include -I/usr/include/python3.10 -c /usr/local/lib/python3.10/dist-packages/mujoco_py/cymj.c -o /usr/local/lib/python3.10/dist-packages/mujoco_py/g

In [ ]:
env = gym.make('CustomHopper-source-v0')
# env = gym.make('CustomHopper-target-v0')
print('State space:', env.observation_space)  # state-space
print('Action space:', env.action_space)  # action-space
print("Mass values of each link:", env.model.body_mass)
print('Dynamics parameters:', env.get_parameters())  # masses of each link of the Hopper
print("Bodies defined in the environment:", env.model.body_names)
print("Number of degrees of freedom (DoFs) of the robot:", env.model.nv)
print("Number of DoFs for each body:", env.model.body_dofnum)
print("Number of actuators:", env.model.nu)

State space: Box(-inf, inf, (11,), float64)
Action space: Box(-1.0, 1.0, (3,), float32)
Mass values of each link: [0.         2.53429174 3.92699082 2.71433605 5.0893801 ]
Dynamics parameters: [2.53429174 3.92699082 2.71433605 5.0893801 ]
Bodies defined in the environment: ('world', 'torso', 'thigh', 'leg', 'foot')
Number of degrees of freedom (DoFs) of the robot: 6
Number of DoFs for each body: [0 3 1 1 1]
Number of actuators: 3


# Analysis of the Hopper Environment in MuJoCo

This document provides a detailed analysis of the Hopper environment based on the provided terminal output and information from the MuJoCo and Gym documentation.

---

## **Question 1.1: What is the state space in the Hopper environment? Is it discrete or continuous?**

### **Answer:**
- **State Space Description**:
  The state space is represented by a **Box** object with shape `(11,)`, meaning it is a vector of 11 continuous values. These values typically include the positions, velocities, and potentially other sensory data (like contact forces) of the Hopper's components.
  
- **Nature of State Space**:
  The state space is **continuous**, as indicated by the range `(-inf, inf)` and the use of the `Box` object.

---

## **Question 1.2: What is the action space in the Hopper environment? Is it discrete or continuous?**

### **Answer:**
- **Action Space Description**:
  The action space is represented by a **Box** object with shape `(3,)`, meaning it consists of 3 continuous values. These correspond to the torques applied to the actuators controlling the Hopper's joints.

- **Nature of Action Space**:
  The action space is **continuous**, as indicated by the range `(-1.0, 1.0)` and the use of the `Box` object.

---

## **Question 1.3: What is the mass value of each link of the Hopper environment, in the source and target variants respectively?**

### **Answer:**
- **Mass Values for the Source Variant**:
  From the terminal output:
Mass values of each link: [0. 2.53429174 3.92699082 2.71433605 5.0893801 ]
These correspond to the masses of the bodies:
- `world`: 0.0 (fixed reference point)
- `torso`: 2.5343
- `thigh`: 3.9270
- `leg`: 2.7143
- `foot`: 5.0894

- **Mass Values for the Target Variant**:
The mass values for the target variant can be obtained by switching the environment initialization to:
```python
env = gym.make('CustomHopper-target-v0')
(Ensure to re-run the relevant command to print the mass values.)

Comparison of Source and Target Variants: Any differences in mass values between the source and target variants must be explicitly checked in the respective initialization. These differences typically simulate dynamics variability to test robustness.

Additional Information Derived from the Environment
Bodies Defined in the Environment:

('world', 'torso', 'thigh', 'leg', 'foot')
These represent the main components of the Hopper system.

Number of Degrees of Freedom (DoFs) of the Robot: 6
This includes translational and rotational movements of the robot.

Number of DoFs for Each Body:
[0 3 1 1 1]
world: 0 (fixed body)
torso: 3
thigh: 1
leg: 1
foot: 1

Number of Actuators:
The Hopper has 3 actuators controlling its joints.

Conclusion

The Hopper environment features a continuous state and action space, making it well-suited for reinforcement learning tasks. Understanding the dynamics, including body masses and degrees of freedom, is crucial for designing robust controllers and algorithms.

In [ ]:
n_episodes = 5

for episode in range(n_episodes):
  done = False
  observation = env.reset()	 # Reset environment to initial state

  while not done:  # Until the episode is over

    action = env.action_space.sample()	# Sample random action

    observation, reward, done, info = env.step(action)	# Step the simulator to the next timestep

**TASK 2**

In [26]:
! python  project-sim2real-rialti-giunti-gjinaj/trainTest.py --model_name sac_hopper --total_timesteps 5000

/usr/local/lib/python3.10/dist-packages/Cython/Distutils/old_build_ext.py:15: DeprecationWarning: dep_util is Deprecated. Use functions from setuptools instead.
  from distutils.dep_util import newer, newer_group
<frozen importlib._bootstrap>:283: DeprecationWarning: the load_module() method is deprecated and slated for removal in Python 3.12; use exec_module() instead
2025-01-10 15:05:48.359757: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2025-01-10 15:05:48.385158: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2025-01-10 15:05:48.392534: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been regi